src/EditorFeatures/Core/IntelliSense/AsyncCompletion/CompletionSource.cs

```diff
@@ -346,24 +346,21 @@ internal sealed class CompletionSource : IAsyncExpandingCompletionSource
                 AsyncCompletionLogger.LogSourceGetContextTicksDataPoint((int)totalStopWatch.Elapsed.TotalMilliseconds, isCanceled: cancellationToken.IsCancellationRequested);
             }
 
-            static VSCompletionContext CombineCompletionContext(VSCompletionContext context1, VSCompletionContext context2)
+            static VSCompletionContext CombineCompletionContext(IAsyncCompletionSession session, VSCompletionContext context1, VSCompletionContext context2)
             {
-                if (context1.Items.IsEmpty && context1.SuggestionItemOptions is null)
+                if (context1.ItemList.IsEmpty && context1.SuggestionItemOptions is null)
                     return context2;
 
-                if (context2.Items.IsEmpty && context2.SuggestionItemOptions is null)
+                if (context2.ItemList.IsEmpty && context2.SuggestionItemOptions is null)
                     return context1;
 
-                using var _ = ArrayBuilder<VSCompletionItem>.GetInstance(context1.Items.Length + context2.Items.Length, out var itemsBuilder);
-                itemsBuilder.AddRange(context1.Items);
-                itemsBuilder.AddRange(context2.Items);
-
+                var completionList = session.CreateCompletionList(context1.ItemList.Concat(context2.ItemList));
```

Comment:
is it not possible to do CreateCompletionList, and then add both arrays to it in two AddRange calls?

Reply:
https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_search?action=contents&text=pooledreadonlylist&type=code&lp=code-Project&filters=ProjectFilters%7BDevDiv%7DRepositoryFilters%7BVS-Platform%7D&pageSize=25&result=DefaultCollection/DevDiv/VS-Platform/GBmain//src/Editor/Core/Def/PooledObjects/PooledReadOnlyList.cs

Reply:
btw, I'm gonna propose a changeto Editor to address yours and Sam's concern.

Reply:
please file a bug to track this addition, @genlu